[GStreamer][EME][Thunder] Add PlayReady support

calvaris opened the pull request:
#### 38fc0ce11b211e73e15546de9b6db7721a83bf9d
<pre>
[GStreamer][EME][Thunder] Add PlayReady support
<a href="https://bugs.webkit.org/show_bug.cgi?id=251587">https://bugs.webkit.org/show_bug.cgi?id=251587</a>

Reviewed by Philippe Normand.

Enable PlayReady if the Thunder plugins are present.

Also add a small check just in case there is no Thunder running in debug mode.

Added key ID corner case hack. Sometimes PlayReady key IDs come with some bytes swapped so we add both original and
swapped key IDs to the hashmap.

* Source/WebCore/platform/graphics/gstreamer/eme/CDMThunder.cpp:
(WebCore::CDMFactoryThunder::supportedKeySystems const):
(WebCore::CDMInstanceSessionThunder::keyUpdatedCallback):

Canonical link: <a href="https://commits.webkit.org/259890@main">https://commits.webkit.org/259890@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc47bd6df8674b48e2abeeb97d6bd8d8e7b48099

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105857 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14925 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38705 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115055 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175181 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109759 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16344 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6219 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98090 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114825 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111606 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12442 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95412 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39983 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94303 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27055 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81651 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8196 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28408 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8678 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5060 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14305 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47954 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6835 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10234 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/dc47bd6df8674b48e2abeeb97d6bd8d8e7b48099. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/9520

philn commented on Source/WebCore/platform/graphics/gstreamer/eme/CDMThunder.cpp:
```diff
@@ -440,6 +446,20 @@ void CDMInstanceSessionThunder::keyUpdatedCallback(KeyIDType&& keyID)
     m_doesKeyStoreNeedMerging |= m_keyStore.add(KeyHandle::create(keyStatus, WTFMove(keyID), BoxPtr<OpenCDMSession>(m_session)));
+    auto instance = cdmInstanceThunder();
+    if (instance && GStreamerEMEUtilities::isPlayReadyKeySystem(instance->keySystem())) {
+        // PlayReady corner case hack: It happens that the key ID
```
Can this hack be conditional to `isThunderRanked()` please?

calvaris replied:
I don't think it is a good idea because you could still have thunder running but not necessarily upranked. Anyway, I think conditional compilation for `THUNDER` would be a good idea.

calvaris replied:
And it is already guarded as this is the _CDMThunder.cpp_ file.

philn replied:
> Can this hack be conditional to `isThunderRanked()` please?

Actually this function doesn't make much sense anymore since the ClearKey-specific decryptor was removed. If you want to handle that encryption format you'd need to implement a Thunder thing for it. No need to play with element ranks :)

philn replied:
There's no code yet for detecting a sparkle-cdm runtime, so you can keep this hack un-guarded. Probably will need a revisit in case a sparkle-cdm playready plugin appears...
Would be nice though to clean out the `isThunderRanked()` thing, can you do it?

calvaris replied:
Can we do it in a follow up?

philn commented on Source/WebCore/platform/graphics/gstreamer/eme/CDMThunder.cpp:
```diff
@@ -440,6 +446,20 @@ void CDMInstanceSessionThunder::keyUpdatedCallback(KeyIDType&& keyID)
+        std::swap(swappedKeyID[4], swappedKeyID[5]);
+        std::swap(swappedKeyID[6], swappedKeyID[7]);
+        GST_MEMDUMP("updated swapped key", swappedKeyID.data(), swappedKeyID.size());
+        m_doesKeyStoreNeedMerging |= m_keyStore.add(KeyHandle::create(keyStatus, WTFMove(swappedKeyID), BoxPtr<OpenCDMSession>(m_session)));
```
I still don't understand why the "swapped" key and the key are both stored in keystore. Can you explain?

calvaris replied:
Because from what've seen, either the original or the swapped can be retrieved depending on the asset.

calvaris committed:
[GStreamer][EME][Thunder] Add PlayReady support
https://bugs.webkit.org/show_bug.cgi?id=251587

Reviewed by Philippe Normand.

Enable PlayReady if the Thunder plugins are present.

Also add a small check just in case there is no Thunder running in debug mode.

Added key ID corner case hack. Sometimes PlayReady key IDs come with some bytes swapped so we add both original and
swapped key IDs to the hashmap.

* Source/WebCore/platform/graphics/gstreamer/eme/CDMThunder.cpp:
(WebCore::CDMFactoryThunder::supportedKeySystems const):
(WebCore::CDMInstanceSessionThunder::keyUpdatedCallback):

Canonical link: https://commits.webkit.org/259890@main

webkit-commit-queue commented:
Committed 259890@main (38fc0ce11b21): <https://commits.webkit.org/259890@main>

Reviewed commits have been landed. Closing PR #9520 and removing active labels.
